File: 101.对称二叉树.py
#
# @lc app=leetcode.cn id=101 lang=python3
#
# [101] 对称二叉树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import collections
class Solution:
    # def isSymmetric(self, root: TreeNode) -> bool:
    #     if root is None:
    #         return True
    #     def recursive(p: TreeNode, q: TreeNode):
    #         if (p is None) ^ (q is None):
    #             return False
    #         if (p is None) and (q is None):
    #             return True
    #         return p.val == q.val and recursive(p.left, q.right) and recursive(p.right, q.left)
    #     return recursive(root.left, root.right)


    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True
        queue = collections.deque()
        queue.append(root.left)
        queue.append(root.right)

        while len(queue):
            tmp1 = queue.popleft()
            tmp2 = queue.popleft()
            if (tmp1 is None) ^ (tmp2 is None):
                return False
            if tmp1 == tmp2 == None:
                continue
            if tmp1.val != tmp2.val:
                return False
            else:
                queue.append(tmp1.left)
                queue.append(tmp2.right)
                queue.append(tmp2.left)
                queue.append(tmp1.right)
        return True
        
            

        
# @lc code=end

File: 101.对称二叉树.cpp
/*
 * @lc app=leetcode.cn id=101 lang=cpp
 *
 * [101] 对称二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include<queue>
using namespace std;

// struct TreeNode {
//     int val;
//     TreeNode *left;
//     TreeNode *right;
//     TreeNode() : val(0), left(nullptr), right(nullptr) {}
//     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
//     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
// };

class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        // return recursive(root, root);
        TreeNode *l, *r;
        queue<TreeNode*> myQueue;
        myQueue.push(root);
        myQueue.push(root);
        while(!myQueue.empty()){
            l = myQueue.front();
            myQueue.pop();
            r = myQueue.front();
            myQueue.pop();
            if(!l ^ !r) return false;
            if(!l && !r) continue;
            if(l->val != r->val) return false;
            myQueue.push(l->left);
            myQueue.push(r->right);
            myQueue.push(l->right);
            myQueue.push(r->left);
        }
        return true;
        
    }

    // bool recursive(TreeNode* left, TreeNode* right){
    //     if(!left && !right) return true;
    //     if(!left || !right) return false;
    //     return left->val == right->val && recursive(left->left, right->right) && recursive(left->right, right->left);
    // }

};
// @lc code=end

